spacy/tokens/_retokenize.pyx

```diff
@@ -47,87 +48,137 @@ cdef class Retokenizer:
-    cdef TokenC* token = &doc.c[start]
-    token.spacy = doc.c[end-1].spacy
-    for attr_name, attr_value in attributes.items():
-        if attr_name == TAG:
-            doc.vocab.morphology.assign_tag(token, attr_value)
-        else:
-            Token.set_struct_attr(token, attr_name, attr_value)
-    # Make sure ent_iob remains consistent
-    if doc.c[end].ent_iob == 1 and token.ent_iob in (0, 2):
-        if token.ent_type == doc.c[end].ent_type:
-            token.ent_iob = 3
-        else:
-            # If they're not the same entity type, let them be two entities
-            doc.c[end].ent_iob = 3
-    # Begin by setting all the head indices to absolute token positions
-    # This is easier to work with for now than the offsets
-    # Before thinking of something simpler, beware the case where a
-    # dependency bridges over the entity. Here the alignment of the
-    # tokens changes.
-    span_root = span.root.i
-    token.dep = span.root.dep
-    # We update token.lex after keeping span root and dep, since
-    # setting token.lex will change span.start and span.end properties
-    # as it modifies the character offsets in the doc
-    token.lex = lex
-    for i in range(doc.length):
-        doc.c[i].head += i
-    # Set the head of the merged token, and its dep relation, from the Span
-    token.head = doc.c[span_root].head
-    # Adjust deps before shrinking tokens
-    # Tokens which point into the merged token should now point to it
-    # Subtract the offset from all tokens which point to >= end
-    offset = (end - start) - 1
-    for i in range(doc.length):
-        head_idx = doc.c[i].head
-        if start <= head_idx < end:
-            doc.c[i].head = start
-        elif head_idx >= end:
-            doc.c[i].head -= offset
-    # Now compress the token array
-    for i in range(end, doc.length):
-        doc.c[i - offset] = doc.c[i]
-    for i in range(doc.length - offset, doc.length):
-        memset(&doc.c[i], 0, sizeof(TokenC))
-        doc.c[i].lex = &EMPTY_LEXEME
-    doc.length -= offset
-    for i in range(doc.length):
+    cdef Span span
+    cdef const LexemeC* lex
+    cdef TokenC* token
+    spans = []
+    cdef TokenC** tokens = <TokenC **>malloc(len(merges) * sizeof(TokenC))
+
+    try:
+        for merge_index, (span, attributes) in enumerate(merges):
+            start = span.start
+            end = span.end
+            spans.append(span)
+
+            # House the new merged token where it starts
+            token = &doc.c[start]
+
+            tokens[merge_index] = token
+
+            # Assign attributes
+            for attr_name, attr_value in attributes.items():
+                if attr_name == TAG:
+                    doc.vocab.morphology.assign_tag(token, attr_value)
+                else:
+                    Token.set_struct_attr(token, attr_name, attr_value)
+
+            # Make sure ent_iob remains consistent
+            if doc.c[end].ent_iob == 1 and token.ent_iob in (0, 2):
```

Comment:
I'm a bit worried about this block. If possible we'd like to be making the `ent_iob` consistent at the end.